mantidproject/mantid
# LoadLiveData should adjust bin boundaries to fit all events when in PreserveEvents mode

Fixes #22406.

**Description of work**

When in PreserveEvents mode, EventListeners were creating workspaces that had incorrect bin boundaries in their X values. This resulted in some algorithms not being able to work with the workspaces as expected, inconsistent behaviour in the Instrument View, etc.

This PR resolves this by introducing a `void IEventWorkspace::updateAllX(void)` method, which replaces HistogramX for all spectra with a single bin that has bin boundaries which are set to the minimum and maximum TOF based on all events. This method is then called in LoadLiveData every time new data is processed.

This occurs after the pre-processing step and overwrites any user-defined rebinning that occurred there. If a user wants custom binning on live data output, this should be done as a post-processing step instead.

**To test:**

Run `FakeISISEventDAE` with default arguments from sidebar.

Then execute the following in the console:
```python
>>> ConfigService.setFacility('TEST_LIVE')
>>> StartLiveData(Instrument="ISIS_Event", OutputWorkspace="test", UpdateEvery=1, PreserveEvents=True, AccumulationMethod="Replace")
```
Open the Instrument View and observe the Time-of-flight range on the bottom, or view the workspace data and observe the X values.

@matthew-d-jones Maximum event throughput performance should also be tested since we saw it dip very low while trying solutions for this. It looks fine to me now, but maybe you can double check.

**Release Notes** 

Release notes have been updated.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
